File: day4/ruby/part_2.rb
def count_xmas(grid)
  ret = 0

  def check_xmas_x_shape(grid, x, y)
    return false if grid[x][y] != 'A'

    top_left = grid[x-1][y-1] + grid[x+1][y+1]
    top_right = grid[x-1][y+1] + grid[x+1][y-1]

    (top_left == "MS" || top_left == "SM") &&
    (top_right == "MS" || top_right == "SM")
  end

  (1..grid.length-2).each do |x|
    (1..grid[0].length-2).each do |y|
      ret += 1 if check_xmas_x_shape(grid, x, y)
    end
  end

  ret
end

grid = File.readlines("input.txt").map(&:strip).map(&:chars)
puts count_xmas(grid)

File: day4/java/Part2.java
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;

public class Part2 {
    private static int countXmas(char[][] grid) {
        int ret = 0;

        for (int x = 1; x < grid.length - 1; x++) {
            for (int y = 1; y < grid[0].length - 1; y++) {
                if (checkXmasXShape(grid, x, y)) {
                    ret++;
                }
            }
        }

        return ret;
    }

    private static boolean checkXmasXShape(char[][] grid, int x, int y) {
        if (grid[x][y] != 'A') {
            return false;
        }

        String topLeft = "" + grid[x-1][y-1] + grid[x+1][y+1];
        String topRight = "" + grid[x-1][y+1] + grid[x+1][y-1];

        return (topLeft.equals("MS") || topLeft.equals("SM")) &&
               (topRight.equals("MS") || topRight.equals("SM"));
    }

    public static void main(String[] args) throws Exception {
        List<String> lines = Files.readAllLines(Paths.get("input.txt"));
        char[][] grid = new char[lines.size()][];
        for (int i = 0; i < lines.size(); i++) {
            grid[i] = lines.get(i).toCharArray();
        }

        System.out.println(countXmas(grid));
    }
}
